File: server/registry.ts
/**
 * JARVIS-OS — Master Agent, Skills, and Tools Registry
 *
 * Provides typed access to all available agents, sub-agents, modular skills,
 * and native tool actuators in the JARVIS-OS ecosystem.
 *
 * Backed by `agents.registry.json`.
 */

import fs from "fs";
import path from "path";

// ── Types ───────────────────────────────────────────────────────

export interface AgentRegistryEntry {
  id: string;
  name: string;
  role: string;
  description: string;
  platform: "gemini-live" | "hermes" | "prime-agent" | "ultron" | "rust-memory-engine" | "openclaw" | "mcp";
  profile?: string;
  thread?: string;
  vaultDirectory?: string;
  status: "online" | "ready" | "offline" | "disabled";
  capabilities: string[];
}

export interface SkillRegistryEntry {
  name: string;
  displayName: string;
  category: "Weather" | "News" | "Productivity" | "Utility" | "System";
  description: string;
  parameters: Record<string, {
    type: string;
    description: string;
    enum?: string[];
  }>;
  executionMode: "sync" | "async-parallel";
  source: string;
}

export interface ToolRegistryEntry {
  binary?: string;
  description: string;
  category?: string;
  status?: string;
  port?: number;
  socketPath?: string;
  source?: string;
}

export interface MasterRegistry {
  version: string;
  name: string;
  description: string;
  updatedAt: string;
  agents: Record<string, AgentRegistryEntry>;
  skills: Record<string, SkillRegistryEntry>;
  tools: {
    cpp_workers: {
      category: string;
      binDirectory: string;
      totalWorkers: number;
      items: Record<string, ToolRegistryEntry>;
    };
    audio_gateway: ToolRegistryEntry;
    memory_engine: ToolRegistryEntry;
    telegram_notifier: ToolRegistryEntry;
    heartbeat_engine: ToolRegistryEntry;
    capability_forge: ToolRegistryEntry;
    [key: string]: any;
  };
}

// ── Singleton Loader ────────────────────────────────────────────

let cachedRegistry: MasterRegistry | null = null;

function loadRegistry(): MasterRegistry {
  if (cachedRegistry) return cachedRegistry;

  const registryPath = path.resolve(process.cwd(), "agents.registry.json");
  if (!fs.existsSync(registryPath)) {
    throw new Error(`Master registry file not found at: ${registryPath}`);
  }

  const raw = fs.readFileSync(registryPath, "utf-8");
  cachedRegistry = JSON.parse(raw) as MasterRegistry;
  return cachedRegistry;
}

// ── Query APIs ──────────────────────────────────────────────────

/**
 * Get the full master registry.
 */
export function getMasterRegistry(): MasterRegistry {
  return loadRegistry();
}

/**
 * Get all registered agents as a list.
 */
export function getRegisteredAgents(): AgentRegistryEntry[] {
  const reg = loadRegistry();
  return Object.values(reg.agents);
}

/**
 * Get a specific agent by its ID (e.g. 'prime-agent', 'jarvis-prime', 'hermes', 'ultron').
 */
export function getAgentById(id: string): AgentRegistryEntry | undefined {
  const reg = loadRegistry();
  return reg.agents[id];
}

/**
 * Get all registered skills as a list.
 */
export function getRegisteredSkills(): SkillRegistryEntry[] {
  const reg = loadRegistry();
  return Object.values(reg.skills);
}

/**
 * Get a specific skill by its name (e.g. 'get_weather_forecast', 'manage_reminders').
 */
export function getSkillByName(name: string): SkillRegistryEntry | undefined {
  const reg = loadRegistry();
  return reg.skills[name];
}

/**
 * Get all native C++ workers and actuators.
 */
export function getRegisteredTools() {
  const reg = loadRegistry();
  return {
    cppWorkers: reg.tools.cpp_workers.items,
    audioGateway: reg.tools.audio_gateway,
    memoryEngine: reg.tools.memory_engine,
    telegramNotifier: reg.tools.telegram_notifier,
    heartbeatEngine: reg.tools.heartbeat_engine,
    capabilityForge: reg.tools.capability_forge,
  };
}

/**
 * Summary metrics for quick system telemetry.
 */
export function getRegistryStats() {
  const reg = loadRegistry();
  return {
    totalAgents: Object.keys(reg.agents).length,
    totalSkills: Object.keys(reg.skills).length,
    totalCppWorkers: Object.keys(reg.tools.cpp_workers.items).length,
    version: reg.version,
    updatedAt: reg.updatedAt,
  };
}

/**
 * Formats a condensed, markdown representation of all available Agents, Skills,
 * and Tools for injection into JARVIS's LLM system prompt context.
 *
 * This allows JARVIS to always be self-aware of what tools and sub-agents it can invoke.
 */
export function generateSystemPromptRegistry(): string {
  const reg = loadRegistry();

  const agentLines = Object.values(reg.agents)
    .map((a) => `- **${a.name}** (\`${a.id}\`): ${a.role} [${a.capabilities.join(", ")}]`)
    .join("\n");

  const skillLines = Object.values(reg.skills)
    .map((s) => `- \`${s.name}\` (${s.category}): ${s.description}`)
    .join("\n");

  const cppWorkerCount = Object.keys(reg.tools.cpp_workers.items).length;

  return `
### 🛠️ JARVIS-OS Capabilities & Sub-Agent Registry
**Delegatable Sub-Agents & Departments:**
${agentLines}

**Available Modular Skills:**
${skillLines}

**Hardware & System Actuators:**
- ${cppWorkerCount} compiled C++ native workers in \`workers_cpp/bin/\` (telemetry, processes, thermals, storage, wifi, desktop automation).
- Rust Audio Gateway (\`/tmp/jarvis_audio.sock\`) for standalone mic/speaker streaming.
- Rust Memory Engine (port 50051) with SQLite WAL & FTS5 Knowledge Graph.
- Capability Forge (\`bwrap\` sandbox) for dynamically synthesizing custom tools.
- Proactive Telegram Notifier & 24/7 Heartbeat.
`.trim();
}

File: brain/registry.py
"""
JARVIS-OS Master Registry Loader for Python Core Engine.
Reads agents.registry.json to provide agent, skill, and tool introspection.
"""

import os
import json
from pathlib import Path
from typing import Dict, Any, List, Optional

REGISTRY_PATH = os.path.join(os.getcwd(), "agents.registry.json")


class MasterRegistry:
    _instance = None

    @classmethod
    def get_instance(cls) -> "MasterRegistry":
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance

    def __init__(self, path: str = REGISTRY_PATH):
        self.path = path
        self._data: Dict[str, Any] = {}
        self.reload()

    def reload(self):
        if os.path.exists(self.path):
            with open(self.path, "r", encoding="utf-8") as f:
                self._data = json.load(f)
        else:
            self._data = {"agents": {}, "skills": {}, "tools": {}}

    @property
    def raw(self) -> Dict[str, Any]:
        return self._data

    def get_agents(self) -> List[Dict[str, Any]]:
        return list(self._data.get("agents", {}).values())

    def get_agent(self, agent_id: str) -> Optional[Dict[str, Any]]:
        return self._data.get("agents", {}).get(agent_id)

    def get_skills(self) -> List[Dict[str, Any]]:
        return list(self._data.get("skills", {}).values())

    def get_skill(self, skill_name: str) -> Optional[Dict[str, Any]]:
        return self._data.get("skills", {}).get(skill_name)

    def get_tools(self) -> Dict[str, Any]:
        return self._data.get("tools", {})

    def get_prompt_summary(self) -> str:
        """Returns a condensed markdown summary for injecting into Jinja2 prompt engine."""
        agents = self.get_agents()
        skills = self.get_skills()
        tools = self.get_tools()

        agent_lines = "\n".join(
            f"- **{a.get('name')}** (`{a.get('id')}`): {a.get('role')} [{', '.join(a.get('capabilities', []))}]"
            for a in agents
        )

        skill_lines = "\n".join(
            f"- `{s.get('name')}` ({s.get('category')}): {s.get('description')}"
            for s in skills
        )

        cpp_items = tools.get("cpp_workers", {}).get("items", {})

        return (
            "### 🛠️ JARVIS-OS Registered Capabilities\n"
            f"**Agents & Departments ({len(agents)}):**\n{agent_lines}\n\n"
            f"**Modular Skills ({len(skills)}):**\n{skill_lines}\n\n"
            f"**Hardware Workers:** {len(cpp_items)} compiled C++ binaries in `workers_cpp/bin/`.\n"
            "- Rust Audio Gateway (`/tmp/jarvis_audio.sock`)\n"
            "- Rust Memory Engine (`port 50051`)\n"
            "- Capability Forge (`bwrap` Sandbox)"
        )


master_registry = MasterRegistry.get_instance()
